### PR TITLE
`azurerm_databricks_workspace` - Upgrade Databricks API to the 2021-04-01-preview version

### LINKED ISSUE #7799: Enhancement: azurerm_databricks_workspace - Add DBFS root Blob storage URL Attribute
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
When implementing private Databricks VNET, you are required to [add User-defined Routes](https://docs.microsoft.com/en-us/azure/databricks/administration-guide/cloud-configurations/azure/udr) to ensure that network traffic is routed correctly for your workspace. 

In the provided link, all of those DNS entries are constant except the DBFS DNS, which is generated at workspace creation.

Here is what I'm using to generate the Route Table and associate it to the Databricks subnets.
```hcl
# https://docs.microsoft.com/en-us/azure/databricks/administration-guide/cloud-configurations/azure/udr
# https://www.terraform.io/docs/providers/dns/d/dns_a_record_set.html
data "dns_a_record_set" "metastore" {
  host = "consolidated-northcentralus-prod-metastore.mysql.database.azure.com"
}

# https://www.terraform.io/docs/providers/dns/d/dns_a_record_set.html
data "dns_a_record_set" "artifact_blob_storage_primary" {
  host = "dbartifactsprodncus.blob.core.windows.net"
}

# https://www.terraform.io/docs/providers/dns/d/dns_a_record_set.html
data "dns_a_record_set" "log_blob_storage" {
  host = "dblogprodncentralus.blob.core.windows.net"
}

# https://www.terraform.io/docs/providers/dns/d/dns_a_record_set.html
data "dns_a_record_set" "event_hub_endpoint" {
  host = "prod-westus-observabilityEventHubs.servicebus.windows.net"
}

# https://www.terraform.io/docs/providers/dns/d/dns_a_record_set.html
data "dns_a_record_set" "dbfs_blob_storage" {
  # https://docs.microsoft.com/en-us/azure/databricks/administration-guide/cloud-configurations/azure/udr#dbfs-root-blob-storage-ip-address
  host = "dbstoragexxxxxxxblob.core.windows.net" # TODO: Need a way to capture this from Databricks resource
  # host = module.databricks_default.default.dbfs_root_url
}

# https://www.terraform.io/docs/providers/azurerm/r/route_table.html
# https://docs.microsoft.com/en-us/azure/databricks/administration-guide/cloud-configurations/azure/on-prem-network#--step-3-create-user-defined-routes-and-associate-them-with-your-azure-databricks-virtual-network-subnets
# https://docs.microsoft.com/en-us/azure/databricks/administration-guide/cloud-configurations/azure/udr
resource "azurerm_route_table" "databricks" {
  name                          = lower(join("-", ["rt", var.NAME, "adb", local.metadata.stage, local.metadata.location]))
  location                      = data.azurerm_resource_group.default.location
  resource_group_name           = data.azurerm_resource_group.default.name
  disable_bgp_route_propagation = true

  route {
    name           = "ControlPlaneNAT"
    address_prefix = "23.101.152.95/32"
    next_hop_type  = "Internet"
  }

  route {
    name           = "Webapp"
    address_prefix = "40.70.58.221/32"
    next_hop_type  = "Internet"
  }

  route {
    name           = "Metastore"
    address_prefix = join("/", [data.dns_a_record_set.metastore.addrs[0], "32"])
    next_hop_type  = "Internet"
  }

  route {
    name           = "ArtifactBlobStorage"
    address_prefix = join("/", [data.dns_a_record_set.artifact_blob_storage_primary.addrs[0], "32"])
    next_hop_type  = "Internet"
  }

  route {
    name           = "LogBlobStorage"
    address_prefix = join("/", [data.dns_a_record_set.log_blob_storage.addrs[0], "32"])
    next_hop_type  = "Internet"
  }

  route {
    name           = "DBFSRootBlobStorage"
    address_prefix = join("/", [data.dns_a_record_set.dbfs_root_blob_storage.addrs[0], "32"])
    next_hop_type  = "Internet"
  }
}

# https://www.terraform.io/docs/providers/azurerm/r/subnet_route_table_association.html
resource "azurerm_subnet_route_table_association" "databricks_private" {
  subnet_id      = azurerm_subnet.databricks_private.id
  route_table_id = azurerm_route_table.databricks.id
}

# https://www.terraform.io/docs/providers/azurerm/r/subnet_route_table_association.html
resource "azurerm_subnet_route_table_association" "databricks_public" {
  subnet_id      = azurerm_subnet.databricks_public.id
  route_table_id = azurerm_route_table.databricks.id
}
```

### New or Affected Resource(s)

* azurerm_databricks_workspace

### Potential Terraform Configuration

```hcl
output "dbfs_root_url" {
  value = azurerm_databricks_workspace.default.dbfs_root_url
}
```

### LINKED ISSUE #11158: Databricks: security group cannot be removed from subnet because it has network intent policy
```
Terraform v0.14.9
+ provider registry.terraform.io/hashicorp/azurerm v2.53.0
```

Get this error:
```
Error: removing Network Security Group Association from Subnet "my-sub" (Virtual Network "my-vnet" / Resource Group "my-rg"): network.SubnetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="NetworkSecurityGroupCannotBeRemovedDueToNipOnSubnet" Message="Network security group cannot be removed from subnet /subscriptions/xxx/resourceGroups/my-rg/providers/Microsoft.Network/virtualNetworks/my-vnet/subnets/my-sub because it has network intent policy /subscriptions/xxx/resourceGroups/my-rg/providers/Microsoft.Network/networkIntentPolicies/adb-eastus-npip-xxx applied." Details=[]
```

During destroy of the below:
```
resource "azurerm_databricks_workspace" "dbw_001" {
    name = "${local.name_prefix}-dbw-001"
    resource_group_name = var.inject_into_vnet.resource_group_name
    location = var.inject_into_vnet.location
    sku = var.databricks_sku

    tags = merge(local.common_tags,{"classifier":"databricks"})

    custom_parameters {
        no_public_ip = true
        virtual_network_id = var.inject_into_vnet.virtual_network_id
        public_subnet_name = azurerm_subnet.snet_dbw_cnt_001.name
        private_subnet_name = azurerm_subnet.snet_dbw_hst_001.name
    }
}
```

https://github.com/Azure/azure-quickstart-templates/issues/6670

### PR DESCRIPTION
**Pull request by WodansSon**
**NOTE:** This PR **cannot** be merged until the service team removes the subscription level feature flag from Azure excluding all subscriptions who have not registered the feature flag within their subscription from calling this API version, else it would break all customers currently using the Terraform Databricks resource.

**ADDITIONAL:** Service team has indicated that this version of the API will be callable by all subscriptions on the 26th of July. So I have moved this into the v2.70.0 Release.

**ADDITIONAL ADDITIONAL:** Removing the feature flag for the API is taking longer than expected to roll out and is now estimated to be complete by Friday 7/30/2021. So I have moved this into the v2.71.0 Release.

* **Dependencies:** Upgrade Databricks API to the 2021-04-01-preview version
* **New Data Source:** `azurerm_databricks_workspace_private_endpoint_connection`

## Added fields:
* public_network_access_enabled
* network_security_group_rules_required
* load_balancer_backend_address_pool_id
* nat_gateway_name
* public_ip_name
* storage_account_name
* storage_account_sku_name
* vnet_address_prefix
* public_subnet_network_security_group_association_id
* private_subnet_network_security_group_association_id

## Data Source fields:
* workspace_id
* private_endpoint_id
* `connections` block
   ## `connections` block fields:
   * workspace_private_endpoint_id
   * name
   * status
   * description
   * action_required

## Examples:
* Added an example configuration file on how to correctly configure `Databricks Workspace` with `Private Link Endpoint`.

fixes #11158
fixes #7799

### CODE REVIEW

File: azurerm/internal/services/databricks/databricks_workspace_resource.go [hunk 1]
```diff
@@ -85,25 +85,76 @@ func resourceDatabricksWorkspace() *pluginsdk.Resource {
 				Default:  false,
 			},
 
+			"public_network_access": {
```
Comment (katbyte) [inline azurerm/internal/services/databricks/databricks_workspace_resource.go:88]:
```suggestion
			"public_network_access_enabled": {
```

File: azurerm/internal/services/databricks/databricks_workspace_resource.go [hunk 2]
```diff
@@ -85,25 +85,76 @@ func resourceDatabricksWorkspace() *pluginsdk.Resource {
...
+				Type:     pluginsdk.TypeBool,
+				ForceNew: true,
+				Optional: true,
+				Default:  true,
+			},
+
+			"require_nsg_rules": {
```
Comment (katbyte) [inline azurerm/internal/services/databricks/databricks_workspace_resource.go:95]:
```suggestion
			"require_network_security_group_rules": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: azurerm/internal/services/databricks/databricks_workspace_resource.go [hunk 3]
```diff
@@ -85,25 +85,76 @@ func resourceDatabricksWorkspace() *pluginsdk.Resource {
...
 				Elem: &pluginsdk.Resource{
 					Schema: map[string]*pluginsdk.Schema{
+						"load_balancer_backend_pool_name": {
+							Type:         pluginsdk.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							AtLeastOneOf: workspaceCustomParametersString(),
+						},
+
+						"load_balancer_id": {
```
Comment (katbyte) [inline azurerm/internal/services/databricks/databricks_workspace_resource.go:121]: can you get the name from the id?

File: azurerm/internal/services/databricks/databricks_workspace_resource.go [hunk 4]
```diff
@@ -85,25 +85,76 @@ func resourceDatabricksWorkspace() *pluginsdk.Resource {
...
 			},
 
+			"public_network_access_enabled": {
+				Type:     pluginsdk.TypeBool,
+				ForceNew: true,
+				Optional: true,
+				Default:  true,
+			},
+
+			"require_network_security_group_rules": {
```
Comment (katbyte) [inline azurerm/internal/services/databricks/databricks_workspace_resource.go:95]: should we be consistant with the tense?
```suggestion
			"network_security_group_rules_required": {
```

File: azurerm/internal/services/databricks/databricks_workspace_resource.go [hunk 5]
```diff
@@ -114,20 +165,58 @@ func resourceDatabricksWorkspace() *pluginsdk.Resource {
...
+
+						"storage_account_name": {
+							Type:         pluginsdk.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							Computed:     true,
+							AtLeastOneOf: workspaceCustomParametersString(),
+						},
+
+						"storage_account_sku_name": {
```
Comment (katbyte) [inline azurerm/internal/services/databricks/databricks_workspace_resource.go:205]: can we get this from the storage account or is it auto created from the name above:

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @WodansSon - just a couple comments left inline

File: azurerm/internal/services/databricks/databricks_workspace_resource.go (hunk 3 above)
Comment (WodansSon) [reply azurerm/internal/services/databricks/databricks_workspace_resource.go:121]: No, I cannot get the backend name from the load balancer but I can get the load balancer ID from the backend pool id. 🤔

File: azurerm/internal/services/databricks/databricks_workspace_resource.go (hunk 5 above)
Comment (WodansSon) [reply azurerm/internal/services/databricks/databricks_workspace_resource.go:205]: No, this is auto created by the Databricks RP, it has a default value of standard_GRS I think. This is offered as an override so you can control sku of the auto generated storage account.

File: azurerm/internal/services/databricks/databricks_workspace_resource.go (hunk 4 above)
Comment (WodansSon) [reply azurerm/internal/services/databricks/databricks_workspace_resource.go:95]: Fixed.

File: azurerm/internal/services/databricks/databricks_workspace_resource.go (hunk 3 above)
Comment (WodansSon) [reply azurerm/internal/services/databricks/databricks_workspace_resource.go]: Fixed... I have removed the two custom parameters and replaced it with a single top level parameter.

**Comment by WodansSon** [comment]
image

**Review by katbyte — APPROVED** [review]
Thanks @WodansSon - LGTM 🍰

**Pull request merged**